add gluon-dns-config: a package that enables the next node to be used as dns cache

christf opened the pull request:
as dns server any server in the network can be used. for example an anycast-ip-address where dns is served ie on all gateways.

rubo77 commented:
Can you please add some examples,where this is very useful?

christf commented:
This shortens the dns round-trip times by factor 2 in our network and reduces the amount of small network packages on vpn-links.

christf commented:
We are seeing cache hit rates up to  70 percent with 5000 entries.

rubo77 commented:
Ah in the [README](https://github.com/freifunk-ffm/gluon/blob/67e8fe87b6405cdc53ecf0c7b149a79e7a2c8b61/package/gluon-dns-cache/README.md) you explained it all, nice. 

so 5000 DNS entries could be cached in about 40KB RAM?

christf commented:
5000 records with roughly 90bytes each will take 450kb of ram.

NeoRaider commented on package/gluon-dns-cache/README.md:
```diff
@@ -0,0 +1,39 @@
+Siehe auch:  
+https://wiki.openwrt.org/doc/uci/dhcp  
+http://flux242.blogspot.de/2012/06/dnsmasq-cache-size-tuning.html
+
```
README should be in English, and rather be moved into *docs*.

christf replied:
done

NeoRaider commented on package/gluon-dns-cache/luasrc/lib/gluon/upgrade/820-dns-config:
```diff
@@ -0,0 +1,14 @@
+uci:set('dhcp', dnsmasq, 'server', site.dns.servers)
+uci:set('dhcp', dnsmasq, 'localise_queries', '1')
+uci:set('dhcp', dnsmasq, 'localservice', '0')
+domain = uci:add('dhcp', 'domain')
```
This would add a new section on each upgrade. Unnamed UCI sections should not be used without a good reason.

NeoRaider commented on package/gluon-dns-cache/luasrc/lib/gluon/upgrade/820-dns-config:
```diff
@@ -0,0 +1,14 @@
+uci:set('dhcp', dnsmasq, 'localise_queries', '1')
+uci:set('dhcp', dnsmasq, 'localservice', '0')
+domain = uci:add('dhcp', 'domain')
+uci:set('dhcp', domain, 'name', 'nextnode')
```
This should not be hardcoded.

christf replied:
what specifically do you mean should not be hardcoded?

NeoRaider commented on package/gluon-dns-cache/Makefile:
```diff
@@ -0,0 +1,37 @@
+endef
+
+$(eval $(call BuildPackage,gluon-dns-config))
+
```
This is too trivial, let's just add it to gluon-core. All new site.conf entries can be made optional, and the upgrade scripts can be adjusted to work with that.

christf replied:
done.

NeoRaider commented on package/gluon-dns-cache/check_site.lua:
```diff
@@ -0,0 +1,3 @@
+need_string_match('dns.cacheentries', '^%d$')
+need_string_array('dns.servers')
```
I don't like hardcoding the DNS servers, but I guess there aren't really better options that don't requite much more work...

NeoRaider commented on package/gluon-dns-cache/luasrc/lib/gluon/upgrade/820-dns-config:
```diff
@@ -0,0 +1,14 @@
+uci:set('dhcp', dnsmasq, 'localservice', '0')
+domain = uci:add('dhcp', 'domain')
+uci:set('dhcp', domain, 'name', 'nextnode')
+uci:set('dhcp', domain, 'ip', site.next_node.ip6 )
```
This should be added for IPv4 as well, to avoid breaking the next-node hostname for clients without IPv6 support.

christf replied:
done

NeoRaider commented on package/gluon-dns-cache/check_site.lua:
```diff
@@ -0,0 +1,3 @@
+need_string_match('dns.cacheentries', '^%d$')
```
This is plain wrong, the value should be a number (as you explain in the README).

christf replied:
done

christf commented:
@NeoRaider  I have reworked this to incorporate your feedback. I am not sure about the implementation of the optional fields.

Also to make the nextnode-name configurable I added it to site.conf in next_node.name.

NeoRaider commented on package/gluon-core/luasrc/lib/gluon/upgrade/820-dns-config:
```diff
@@ -0,0 +1,37 @@
+		end
+	end
+	uci:save('dhcp')
+end
```
I think most of these options don't need to depend on *site.dns.servers*. We can just unconditionally set *localise_queries* and *localservice*, and set each of the other options/sections when it is given in *site.conf*.

We would also need to disable *peerdns* for the client interface when *site.dns.servers* is given, to avoid the node picking up the next-node address as upstream DNS.

Please also add code to remove/reset all these options when the corresponding values don't exist in *site.conf* (anymore).

NeoRaider replied:
You will still need to add these options to `check_site.lua` to validate the types of the values. There are many examples for optional values in other packages, marked as optional by passing *false* as last argument to the different check functions.

christf replied:
done.

NeoRaider commented on docs/features/dns-cache.rst:
```diff
@@ -0,0 +1,27 @@
+      name = 'nextnode',
+      ip6 = 'dddd::1',
+      ip4 = '1.2.3.4',
+ }
```
All addresses in this section should use the designated documentation prefixes (192.0.2.0/24, 198.51.100.0/24 and 203.0.113.0/24 [[RFC5737](https://tools.ietf.org/html/rfc5737)]; 2001:db8::/32 [[RFC3849](https://tools.ietf.org/html/rfc3849)]).

christf replied:
done. Thank you for the links.

christf replied:
done.

christf commented:
Thank you for your feedback. Should I squash commits on this one?

NeoRaider commented on package/gluon-core/luasrc/lib/gluon/upgrade/820-dns-config:
```diff
@@ -0,0 +1,48 @@
+end
+uci:save('dhcp')
+
+uci:set('network', 'client', 'peerdns','0')
```
Ah, I meant the local-node interface, not the client interface, we switched these -.-

It must be set to 0 if `site.dns.servers` is set, to 1 otherwise. This should be done in `package/gluon-mesh-batman-adv-core/luasrc/lib/gluon/upgrade/330-gluon-mesh-batman-adv-core-local-node`, where it is set to 1 at the moment.

NeoRaider commented on package/gluon-core/luasrc/lib/gluon/upgrade/820-dns-config:
```diff
@@ -0,0 +1,48 @@
+	end
+else
+	uci:delete('dhcp', 'domain', 'nextnode4')
+	uci:delete('dhcp', 'domain', 'nextnode6')
```
Maybe get rid of the duplicate delete calls by changing the conditions to `site.next_node.name and site.next_node.ip4`, getting rid of the outer level of ifs.

christf commented:
@NeoRaider changes have been made.

NeoRaider commented on package/gluon-core/luasrc/lib/gluon/upgrade/820-dns-config:
```diff
@@ -0,0 +1,40 @@
+uci:set('dhcp', dnsmasq, 'localise_queries', '1')
+uci:set('dhcp', dnsmasq, 'localservice', '0')
+
+if site.dns.servers then
```
This needs to be `if site.dns and site.dns.servers`.

christf replied:
done

NeoRaider commented on package/gluon-core/luasrc/lib/gluon/upgrade/820-dns-config:
```diff
@@ -0,0 +1,40 @@
+	uci:delete('dhcp', dnsmasq, 'server')
+end
+
+if site.dns.cacheentries then
```
Similar.

christf replied:
done

NeoRaider commented on package/gluon-core/luasrc/lib/gluon/upgrade/820-dns-config:
```diff
@@ -0,0 +1,40 @@
+	uci:delete('dhcp', dnsmasq, 'cachesize')
+end
+
+if site.next_node.name and site.next_node.ip4 then
```
Similar.

christf replied:
done

NeoRaider commented on package/gluon-core/luasrc/lib/gluon/upgrade/820-dns-config:
```diff
@@ -0,0 +1,40 @@
+	uci:delete('dhcp', 'domain', 'nextnode4')
+end
+
+if site.next_node.name and site.next_node.ip6 then
```
Similar.

christf replied:
done

NeoRaider commented on package/gluon-mesh-batman-adv-core/luasrc/lib/gluon/upgrade/330-gluon-mesh-batman-adv-core-local-node:
```diff
@@ -29,6 +29,10 @@ uci:section('network', 'interface', 'local_node',
 	}
 )
 
+if site.dns.servers then
```
Similar.

christf replied:
done

NeoRaider replied:
You forgot this one.

NeoRaider commented on package/gluon-core/check_site.lua:
```diff
@@ -39,3 +39,8 @@ for _, config in ipairs({'wifi24', 'wifi5'}) do
 need_boolean('poe_passthrough', false)
+
+need_number('dns.cacheentries', false)
+need_string_array('dns.servers', false)
```
While we're not always using these functions consistently, it would be better to wrap the `dns` checks in a
```lua
if need_table('dns', nil, false) then
```
block (and the same with `next_node`) to ensure that `site.dns` is a table if it exists.

christf replied:
done.
